File: solutions/2406. Divide Intervals Into Minimum Number of Groups/2406.rs
// Time complexity: O(nlogn)
// Space complexity: O(n)
use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn min_groups(intervals: Vec<Vec<i32>>) -> i32 {
        let mut intervals = intervals;
        intervals.sort_unstable();

        let mut min_heap = BinaryHeap::new();

        for interval in intervals {
            if let Some(&Reverse(top)) = min_heap.peek() {
                if interval[0] > top {
                    min_heap.pop();
                }
            }
            min_heap.push(Reverse(interval[1]));
        }

        min_heap.len() as i32
    }
}

File: solutions/2406. Divide Intervals Into Minimum Number of Groups/2406.py
# Time complexity: O(nlogn)
# Space complexity: O(n)
class Solution:
  def minGroups(self, intervals: List[List[int]]) -> int:
    minHeap = []

    for left, right in sorted(intervals):
      if minHeap and left > minHeap[0]:
        heapq.heappop(minHeap)
      heapq.heappush(minHeap, right)

    return len(minHeap)